src/Http/Authentication.Core/src/Microsoft.AspNetCore.Authentication.Core.csproj

```diff
@@ -16,4 +16,8 @@
     <Reference Include="Microsoft.AspNetCore.Http.Extensions" />
   </ItemGroup>
 
+  <ItemGroup>
+    <InternalsVisibleTo Include="Microsoft.AspNetCore" />
```

Comment:
I spoke to @captainsafia about this and I think we can cautiously allow IVT to `Microsoft.AspNetCore` but not between individual components.

Reply:
> Is there another way to do this? Can we use a public service type as our marker?

That's where I originally started, but @davidfowl mentioned that using a marker type would be a more reliable option. In fact, the authorization middleware already does this here:

https://github.com/dotnet/aspnetcore/blob/5c3a312e6c8798efb62b9b5ccd0ea8620be11118/src/Security/Authorization/Policy/src/AuthorizationAppBuilderExtensions.cs#L40-L45

This serves as a more reliable indicator in the event that users do something strange and register a public type in DI themselves (unlikely but ya never know).

We could've made these marker types public, but I think it pollutes the public API primarily for our benefit with little value to users.

Reply:
My concern is that this makes it easier to consume internal APIs from the `Microsoft.AspNetCore` project in the future without realizing it. In many cases consuming an internal API from another project is a good signal it API should be public and after this change we'll lose that signal.

While no one else may ever use a public `AuthenticationMarkerService` in particular, I don't think the noise it adds is too onerous either. We can make it `EditorVisibleNever` like @Tratcher suggests.

Reply:
Using `EditorVisibleNever` is a good recommendation. I'll go ahead and do that on the existing authorization marker type too.

> While no one else may ever use a public AuthenticationMarkerService in particular, I don't think the noise it adds is too onerous either.

Perhaps, but I think it's the _accumulation_ of decisions like these that pollutes an API.

Reply:
We can always put them in their own 'infrastructure' namespace too.

Reply:
> My concern is that this makes it easier to consume internal APIs from the Microsoft.AspNetCore project in the future without realizing it. In many cases consuming an internal API from another project is a good signal it API should be public and after this change we'll lose that signal.

I hear you but in this particular case what does that matter? **Microsoft.AspNetCore** is the reference everything assembly, so what does it matter that it's IVT to assemblies that it references? Doesn't seem like much of a layer violation IMHO. I feel like it opens up the door for some other cleaner (without public API) opportunities for Microsoft.AspNetCore.Hosting as well.

Reply:
> I would rather not use the marker types and use existing public ones then introduce public APIs with low value.

I don't see the problem with this either. It seems to be the simplest solution. I guess in theory someone could add some subset of public services without calling `AddAuthentication`, but why do we care?

Reply:
PUBTERNAL!!!

Reply:
Went ahead and removed the IVT here. IMO, if we're choosing between introducing hidden, public types or using a non-marker service, I lean towards the later.

Reply:
I think there are some very good usages of IVT here now that I’ve seen this.

Reply:
I agree that IVT could be *very* convenient. I've long thought will make the WAB much cleaner if we were not restricted to using the public `ConfigureWebHostDefaults` API or being forced to add a public API that will ever only be used by WAB itself. I can probably be convinced that IVT is justifiable for `Microsoft.AspNetCore` from any assembly because it's the composition root or whatever giving it a special status.

We should discuss this in a meeting with everyone involved though. If we agree, we can go back to the internal marker service. I don't think it should hold up this PR.